pennylane/gradients/finite_difference.py

```diff
@@ -359,3 +359,255 @@ def processing_fn(results):
+        c0 = coeffs[0]
+        shifts = shifts[1:]
+        coeffs = coeffs[1:]
+
+    method_map = choose_grad_methods(diff_methods, argnum)
+
+    for i, _ in enumerate(tape.trainable_params):
+        if i not in method_map or method_map[i] == "0":
+            # parameter has zero gradient
+            shapes.append(0)
+            continue
+
+        g_tapes = generate_shifted_tapes(tape, i, shifts * h)
+        gradient_tapes.extend(g_tapes)
+        shapes.append(len(g_tapes))
+
+    def processing_fn(results):
+        grads = []
+        start = 1 if c0 is not None and f0 is None else 0
+        r0 = f0 or results[0]
+
+        for s in shapes:
+
+            if s == 0:
+                # parameter has zero gradient
+                g = qml.math.zeros_like(results[0])
+                grads.append(g)
+                continue
+
+            res = results[start : start + s]
+            start = start + s
+            # compute the linear combination of results and coefficients
+            # First compute the multiplication with coeff
+            l = []
+            for i, c in enumerate(coeffs):
+                if isinstance(res[i], (tuple, list)):
+                    elem = [r * c for r in res[i]]
+                else:
+                    elem = [res[i] * c]
+                l.append(elem)
+            # Second add all the term for each measurement separately
+            g = []
+
+            for i in range(0, len(tape.measurements)):
+                elem = sum(r[i] for r in l)
+                g.append(elem)
+
+            # Add on the unshifted term
+            if c0 is not None:
+                # unshifted term
+                if isinstance(r0, (tuple, list)):
+                    c0r0 = [c0 * r for r in r0]
+                else:
+                    c0r0 = [c0 * r0]
+                g = [i + j for i, j in zip(g, c0r0)]
+
+            if len(g) > 1:
+                if isinstance(results[0][0], np.ndarray) and len(tape):
+                    grads.append(tuple(np.array(i / (h**n)) for i in g))
+                else:
+                    grads.append(tuple(i / (h**n) for i in g))
+            else:
+                if isinstance(results[0], np.ndarray):
+                    grads.append(np.array(g[0] / (h**n)))
+                else:
+                    grads.append(g[0] / (h**n))
+
+        # Single measurement
+        if len(tape.measurements) == 1:
+            return tuple(elem for elem in grads)
+
+        # Reordering to match the right shape for multiple measurements
+        grads_reorder = [
+            [0 for _ in range(0, len(tape.trainable_params))]
+            for _ in range(0, len(tape.measurements))
+        ]
+        for i in range(0, len(tape.measurements)):
+            for j in range(0, len(tape.trainable_params)):
+                grads_reorder[i][j] = grads[j][i]
+
+        # To tuple
+        grads_tuple = tuple(tuple(elem) for elem in grads_reorder)

```

Comment:
Again, this will be a tuple "all the way down". Do we instead want a tuple of **tensors**, with each entry corresponding to a measurement and each tensor having shape `(num_params, output_dims)`?

Reply:
At this point we have multiple measurements, therefore yes it is a tuple of tuple. For the one measurement case it is returned just before that.